File: LeetCode - Top Interview 150/20.py
# 20. Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/


class Solution:
    def isValid1(self, s: str) -> bool:
        parentheses_pair = {"(": ")", "{": "}", "[": "]"}

        stack = []

        for parenthesis in s:
            if parenthesis in parentheses_pair:
                stack.append(parentheses_pair[parenthesis])

                continue

            last_one = None
            if len(stack) > 0:
                last_one = stack.pop()

            if last_one != parenthesis:
                return False

        if len(stack) != 0:
            return False

        return True

    def isValid2(self, s: str) -> bool:
        parentheses_pair = {"(": ")", "{": "}", "[": "]"}

        stack = []

        for parenthesis in s:
            if parenthesis in parentheses_pair:
                stack.append(parentheses_pair[parenthesis])

                continue

            if len(stack) == 0 or stack.pop() != parenthesis:
                return False

        if len(stack) != 0:
            return False

        return True


solution = Solution()

print(solution.isValid1(s="()"))
print(solution.isValid1(s="()[]{}"))
print(solution.isValid1(s="(]"))
print(solution.isValid1(s="["))
print(solution.isValid1(s="]"))

File: LeetCode - Algorithm/Stack and Queue/20.js
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const openSet = new Set(["(", "{", "["]);
  const closeSet = new Set([")", "}", "]"]);
  const openToClose = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  const stack = [];

  for (let i = 0; i < s.length; i++) {
    const char = s.charAt(i);

    if (openSet.has(char)) stack.push(char);
    else if (closeSet.has(char)) {
      const lastOne = stack.pop();
      if (openToClose[lastOne] !== char) {
        return false;
      }
    }
  }

  if (stack.length > 0) {
    return false;
  }

  return true;
};

/*
s = "()[]{}"
     ^
*/
